### ISSUE TITLE
Training stability without loss scaling

### ISSUE THREAD

**Issue by hwijeen** [allenai/OLMo#524]
### ❓ The question

Hi, thank you for the great project!

I realized that the training code [does not use loss scaling](https://github.com/allenai/OLMo/blob/194012a08c11c25bd7ddb9a3dafb0e80798ea2a8/olmo/train.py#L675) while employing mixed precision training. Although the coupling is not a must, my experience tells me that it will result in an unstable training, especially in 70B+ models. Was this a intended decision? If so, could you share your intuition as to why it would be a better option? Thanks!

**Comment by hwijeen** [comment]
Right after posting I just realized that it might be because the model was trained with `bf16`, not `fp16` 😅

**Comment by athrvkk** [comment]
Hi! It looks like the current implementation works correctly for fp32 and bf16, but does not apply loss scaling when using fp16. As a result, training with fp16 appears to suffer from diverging loss and gradient explosion.
Would simply adding torch.cuda.amp.GradScaler for fp16 training address this issue?
